antrea-controller panics in triggerCNPUpdates

**Describe the bug**
<!--
A clear and concise description of what the bug is.
If you believe this bug is a security issue, please don't use this template and follow our [security guidelines](/SECURITY.md)
-->
`triggerCNPUpdates` panics if a ClusterGroup is used by multiple ClusterNetworkPolicies. It's because:
When a ClusterGroup is created, updated, or deleted, `triggerCNPUpdates` will be called to update ClusterNetworkPolicies that reference to it. However, a ClusterNetworkPolicy may have not been processed by `addCNP` when they are found in the informer cache, which means its internal NetworkPolicy doesn't exist yet.
```
I0907 12:55:15.953155       1 clusternetworkpolicy.go:37] Processing ClusterNetworkPolicy 60e3ec56-f42c-4508-9ad5-dc44b7043b43 ADD event
E0907 12:55:15.955446       1 runtime.go:78] Observed a panic: &runtime.TypeAssertionError{_interface:(*runtime._type)(0x209ef80), concrete:(*runtime._type)(nil), asserted:(*runtime._type)(0x20ae060), missingMethod:""} (interface conversion: interface {} is nil, not *types.NetworkPolicy)
goroutine 537 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x2117ea0, 0xc000d3d620)
        /tmp/gopath/pkg/mod/k8s.io/apimachinery@v0.21.0/pkg/util/runtime/runtime.go:74 +0x95
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /tmp/gopath/pkg/mod/k8s.io/apimachinery@v0.21.0/pkg/util/runtime/runtime.go:48 +0x89
panic(0x2117ea0, 0xc000d3d620)
        /usr/local/go/src/runtime/panic.go:969 +0x175
antrea.io/antrea/pkg/controller/networkpolicy.(*NetworkPolicyController).triggerCNPUpdates(0xc000654c80, 0xc00213fc80, 0x23ec3d7, 0x4)
        /usr/src/antrea.io/antrea/pkg/controller/networkpolicy/clustergroup.go:280 +0x8c5
antrea.io/antrea/pkg/controller/networkpolicy.(*NetworkPolicyController).syncInternalGroup(0xc000654c80, 0xc00280ca20, 0x24, 0x0, 0x0)
        /usr/src/antrea.io/antrea/pkg/controller/networkpolicy/clustergroup.go:241 +0x4a5
antrea.io/antrea/pkg/controller/networkpolicy.(*NetworkPolicyController).processNextInternalGroupWorkItem(0xc000654c80, 0x203000)
        /usr/src/antrea.io/antrea/pkg/controller/networkpolicy/clustergroup.go:187 +0xfb
antrea.io/antrea/pkg/controller/networkpolicy.(*NetworkPolicyController).internalGroupWorker(0xc000654c80)
        /usr/src/antrea.io/antrea/pkg/controller/networkpolicy/clustergroup.go:169 +0x2b
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000e9fb60)
        /tmp/gopath/pkg/mod/k8s.io/apimachinery@v0.21.0/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000e9fb60, 0x26bc460, 0xc000b4fc80, 0x26bd201, 0xc00056a6c0)
        /tmp/gopath/pkg/mod/k8s.io/apimachinery@v0.21.0/pkg/util/wait/wait.go:156 +0xad
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000e9fb60, 0x3b9aca00, 0x0, 0xc00060db01, 0xc00056a6c0)
        /tmp/gopath/pkg/mod/k8s.io/apimachinery@v0.21.0/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.Until(0xc000e9fb60, 0x3b9aca00, 0xc00056a6c0)
        /tmp/gopath/pkg/mod/k8s.io/apimachinery@v0.21.0/pkg/util/wait/wait.go:90 +0x4d
created by antrea.io/antrea/pkg/controller/networkpolicy.(*NetworkPolicyController).Run
        /usr/src/antrea.io/antrea/pkg/controller/networkpolicy/networkpolicy_controller.go:1012 +0x310
panic: interface conversion: interface {} is nil, not *types.NetworkPolicy [recovered]
        panic: interface conversion: interface {} is nil, not *types.NetworkPolicy
```

**To Reproduce**

Create a ClusterGroup and multiple ClusterNetworkPolicy that use it

**Versions:**
 - Antrea version (Docker image tag). 1.2.0

--- a/pkg/controller/networkpolicy/clustergroup.go
+++ b/pkg/controller/networkpolicy/clustergroup.go
@@ -265,42 +265,8 @@ func (c *NetworkPolicyController) triggerCNPUpdates(cg string) {
 		return
 	}
 	for _, obj := range cnps {
-		cnp := obj.(*crdv1alpha1.ClusterNetworkPolicy)
-		// Re-process ClusterNetworkPolicies which may be affected due to updates to CG.
-		curInternalNP := c.processClusterNetworkPolicy(cnp)
-		klog.V(2).Infof("Updating existing internal NetworkPolicy %s for %s", curInternalNP.Name, curInternalNP.SourceRef.ToString())
-		key := internalNetworkPolicyKeyFunc(cnp)
-		// Lock access to internal NetworkPolicy store such that concurrent access
-		// to an internal NetworkPolicy is not allowed. This will avoid the
-		// case in which an Update to an internal NetworkPolicy object may
-		// cause the SpanMeta member to be overridden with stale SpanMeta members
-		// from an older internal NetworkPolicy.
-		c.internalNetworkPolicyMutex.Lock()
-		oldInternalNPObj, _, _ := c.internalNetworkPolicyStore.Get(key)
-		oldInternalNP := oldInternalNPObj.(*antreatypes.NetworkPolicy)
-		// Must preserve old internal NetworkPolicy Spac.
-		curInternalNP.SpanMeta = oldInternalNP.SpanMeta
-		c.internalNetworkPolicyStore.Update(curInternalNP)
-		// Unlock the internal NetworkPolicy store.
-		c.internalNetworkPolicyMutex.Unlock()
-		// Enqueue addressGroup keys to update their group members.
-		// TODO: optimize this to avoid enqueueing address groups when not updated.
-		for _, atg := range curInternalNP.AppliedToGroups {
-			c.enqueueAppliedToGroup(atg)
-		}
-		for _, rule := range curInternalNP.Rules {
-			for _, addrGroupName := range rule.From.AddressGroups {
-				c.enqueueAddressGroup(addrGroupName)
-			}
-			for _, addrGroupName := range rule.To.AddressGroups {
-				c.enqueueAddressGroup(addrGroupName)
-			}
-		}
-		c.enqueueInternalNetworkPolicy(key)
-		c.deleteDereferencedAddressGroups(oldInternalNP)
-		for _, atg := range oldInternalNP.AppliedToGroups {
-			c.deleteDereferencedAppliedToGroup(atg)
-		}
+		// ClusterGroup may be used by AppliedToGroup, enqueuing them after reprocessing CNP.
+		c.reprocessCNP(obj.(*crdv1alpha1.ClusterNetworkPolicy), true)
 	}
 }
 

--- a/pkg/controller/networkpolicy/clustergroup_test.go
+++ b/pkg/controller/networkpolicy/clustergroup_test.go
@@ -18,8 +18,11 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/sets"
 
 	"antrea.io/antrea/pkg/apis/controlplane"
@@ -827,4 +830,153 @@ func TestGetGroupMembers(t *testing.T) {
 			assert.Equal(t, tt.expectedMembers, members)
 		})
 	}
+}
+
+func TestSyncInternalGroup(t *testing.T) {
+	p10 := float64(10)
+	p20 := float64(20)
+	allowAction := crdv1alpha1.RuleActionAllow
+	cgName := "cgA"
+	cgUID := types.UID("uidA")
+	cg := &crdv1alpha3.ClusterGroup{
+		ObjectMeta: metav1.ObjectMeta{Name: cgName, UID: cgUID},
+		Spec:       crdv1alpha3.GroupSpec{NamespaceSelector: &selectorA},
+	}
+	cnp1 := &crdv1alpha1.ClusterNetworkPolicy{
+		ObjectMeta: metav1.ObjectMeta{Name: "cnp1", UID: "uid1"},
+		Spec: crdv1alpha1.ClusterNetworkPolicySpec{
+			AppliedTo: []crdv1alpha1.NetworkPolicyPeer{
+				{PodSelector: &selectorB},
+			},
+			Priority: p10,
+			Ingress: []crdv1alpha1.Rule{
+				{
+					From: []crdv1alpha1.NetworkPolicyPeer{
+						{Group: cgName},
+					},
+					Action: &allowAction,
+				},
+			},
+		},
+	}
+	cnp2 := &crdv1alpha1.ClusterNetworkPolicy{
+		ObjectMeta: metav1.ObjectMeta{Name: "cnp2", UID: "uid2"},
+		Spec: crdv1alpha1.ClusterNetworkPolicySpec{
+			AppliedTo: []crdv1alpha1.NetworkPolicyPeer{
+				{PodSelector: &selectorC},
+			},
+			Priority: p20,
+			Ingress: []crdv1alpha1.Rule{
+				{
+					From: []crdv1alpha1.NetworkPolicyPeer{
+						{Group: cgName},
+					},
+					Action: &allowAction,
+				},
+			},
+		},
+	}
+
+	_, npc := newControllerWithoutEventHandler(nil, []runtime.Object{cnp1, cnp2, cg})
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	npc.crdInformerFactory.Start(stopCh)
+	npc.crdInformerFactory.WaitForCacheSync(stopCh)
+
+	// After creating a ClusterGroup:
+	// - A corresponding internal group should be added for it.
+	// - The internal NetworkPolicies for the ClusterNetworkPolicies that use it should reference to it regardless of
+	//   whether they are added before the ClusterGroup or after it.
+	// - An AddressGroup should be created for it.
+
+	// cnp1 is added before the ClusterGroup.
+	npc.addCNP(cnp1)
+	npc.addClusterGroup(cg)
+	err := npc.syncInternalGroup(internalGroupKeyFunc(cg))
+	require.NoError(t, err)
+	// cnp2 is added after the ClusterGroup.
+	npc.addCNP(cnp2)
+
+	expectedInternalNetworkPolicy1 := &antreatypes.NetworkPolicy{
+		UID:  "uid1",
+		Name: "uid1",
+		SourceRef: &controlplane.NetworkPolicyReference{
+			Type: controlplane.AntreaClusterNetworkPolicy,
+			Name: "cnp1",
+			UID:  "uid1",
+		},
+		Priority:     &p10,
+		TierPriority: &DefaultTierPriority,
+		Rules: []controlplane.NetworkPolicyRule{
+			{
+				Direction: controlplane.DirectionIn,
+				From:      controlplane.NetworkPolicyPeer{AddressGroups: []string{cgName}},
+				Priority:  0,
+				Action:    &allowAction,
+			},
+		},
+		AppliedToGroups: []string{getNormalizedUID(toGroupSelector("", &selectorB, nil, nil).NormalizedName)},
+	}
+	actualInternalNetworkPolicy1, exists, _ := npc.internalNetworkPolicyStore.Get(internalNetworkPolicyKeyFunc(cnp1))
+	require.True(t, exists)
+	assert.Equal(t, expectedInternalNetworkPolicy1, actualInternalNetworkPolicy1)
+
+	expectedInternalNetworkPolicy2 := &antreatypes.NetworkPolicy{
+		UID:  "uid2",
+		Name: "uid2",
+		SourceRef: &controlplane.NetworkPolicyReference{
+			Type: controlplane.AntreaClusterNetworkPolicy,
+			Name: "cnp2",
+			UID:  "uid2",
+		},
+		Priority:     &p20,
+		TierPriority: &DefaultTierPriority,
+		Rules: []controlplane.NetworkPolicyRule{
+			{
+				Direction: controlplane.DirectionIn,
+				From:      controlplane.NetworkPolicyPeer{AddressGroups: []string{cgName}},
+				Priority:  0,
+				Action:    &allowAction,
+			},
+		},
+		AppliedToGroups: []string{getNormalizedUID(toGroupSelector("", &selectorC, nil, nil).NormalizedName)},
+	}
+	actualInternalNetworkPolicy2, exists, _ := npc.internalNetworkPolicyStore.Get(internalNetworkPolicyKeyFunc(cnp2))
+	require.True(t, exists)
+	assert.Equal(t, expectedInternalNetworkPolicy2, actualInternalNetworkPolicy2)
+
+	expectedInternalGroup := &antreatypes.Group{
+		UID:      cgUID,
+		Name:     cgName,
+		Selector: toGroupSelector("", nil, &selectorA, nil),
+	}
+	actualInternalGroup, exists, _ := npc.internalGroupStore.Get(internalGroupKeyFunc(cg))
+	require.True(t, exists)
+	assert.Equal(t, expectedInternalGroup, actualInternalGroup)
+	_, exists, _ = npc.addressGroupStore.Get(cgName)
+	require.True(t, exists, "An AddressGroup should be created for the ClusterGroup when it's referenced by any ClusterNetworkPolicy")
+
+	// After deleting the ClusterGroup:
+	// - Its corresponding internal group should be removed.
+	// - The internal NetworkPolicies for the ClusterNetworkPolicies that use it should be updated.
+	// - The AddressGroup created for it should be deleted.
+	npc.deleteClusterGroup(cg)
+	err = npc.syncInternalGroup(internalGroupKeyFunc(cg))
+	require.NoError(t, err)
+
+	_, exists, _ = npc.internalGroupStore.Get(internalGroupKeyFunc(cg))
+	require.False(t, exists)
+
+	expectedInternalNetworkPolicy1.Rules[0].From.AddressGroups = nil
+	actualInternalNetworkPolicy1, exists, _ = npc.internalNetworkPolicyStore.Get(internalNetworkPolicyKeyFunc(cnp1))
+	require.True(t, exists)
+	assert.Equal(t, expectedInternalNetworkPolicy1, actualInternalNetworkPolicy1)
+
+	expectedInternalNetworkPolicy2.Rules[0].From.AddressGroups = nil
+	actualInternalNetworkPolicy2, exists, _ = npc.internalNetworkPolicyStore.Get(internalNetworkPolicyKeyFunc(cnp2))
+	require.True(t, exists)
+	assert.Equal(t, expectedInternalNetworkPolicy2, actualInternalNetworkPolicy2)
+
+	_, exists, _ = npc.addressGroupStore.Get(cgName)
+	require.False(t, exists, "The AddressGroup for the ClusterGroup should be deleted when it's no longer referenced by any ClusterNetworkPolicy")
 }

--- a/pkg/controller/networkpolicy/clusternetworkpolicy.go
+++ b/pkg/controller/networkpolicy/clusternetworkpolicy.go
@@ -126,12 +126,14 @@ func (n *NetworkPolicyController) deleteCNP(old interface{}) {
 	n.deleteDereferencedAddressGroups(oldInternalNP)
 }
 
-// reprocessCNP is triggered by Namespace ADD/UPDATE/DELETE events when they impact the
-// per-namespace rules of a CNP.
-func (n *NetworkPolicyController) reprocessCNP(cnp *crdv1alpha1.ClusterNetworkPolicy) {
+// reprocessCNP is triggered when a CNP may be impacted by non-ClusterNetworkPolicy events, including Namespace events
+// (for per-namespace rules) and ClusterGroup events (for ClusterGroup reference).
+func (n *NetworkPolicyController) reprocessCNP(cnp *crdv1alpha1.ClusterNetworkPolicy, enqueueAppliedToGroup bool) {
 	key := internalNetworkPolicyKeyFunc(cnp)
 	n.internalNetworkPolicyMutex.Lock()
 	oldInternalNPObj, exist, _ := n.internalNetworkPolicyStore.Get(key)
+	// The internal NetworkPolicy may haven't been created yet. It's fine to skip processing this CNP as addCNP will
+	// create it eventually.
 	if !exist {
 		klog.V(2).Infof("Cannot find the original internal NetworkPolicy, skip reprocessCNP")
 		n.internalNetworkPolicyMutex.Unlock()
@@ -145,6 +147,11 @@ func (n *NetworkPolicyController) reprocessCNP(cnp *crdv1alpha1.ClusterNetworkPo
 	curInternalNP.SpanMeta = oldInternalNP.SpanMeta
 	n.internalNetworkPolicyStore.Update(curInternalNP)
 	n.internalNetworkPolicyMutex.Unlock()
+	if enqueueAppliedToGroup {
+		for _, atg := range curInternalNP.AppliedToGroups {
+			n.enqueueAppliedToGroup(atg)
+		}
+	}
 	// Enqueue addressGroup keys to update their Node span.
 	for _, rule := range curInternalNP.Rules {
 		for _, addrGroupName := range rule.From.AddressGroups {
@@ -196,7 +203,7 @@ func (n *NetworkPolicyController) addNamespace(obj interface{}) {
 	affectedACNPs := n.filterPerNamespaceRuleACNPsByNSLabels(namespace.Labels)
 	for cnpName := range affectedACNPs {
 		if cnp, err := n.cnpLister.Get(cnpName); err == nil {
-			n.reprocessCNP(cnp)
+			n.reprocessCNP(cnp, false)
 		}
 	}
 }
@@ -213,7 +220,7 @@ func (n *NetworkPolicyController) updateNamespace(oldObj, curObj interface{}) {
 	affectedACNPs := utilsets.SymmetricDifferenceString(affectedACNPsByOldLabels, affectedACNPsByCurLabels)
 	for cnpName := range affectedACNPs {
 		if cnp, err := n.cnpLister.Get(cnpName); err == nil {
-			n.reprocessCNP(cnp)
+			n.reprocessCNP(cnp, false)
 		}
 	}
 }
@@ -243,7 +250,7 @@ func (n *NetworkPolicyController) deleteNamespace(old interface{}) {
 			klog.Errorf("Error getting Antrea ClusterNetworkPolicy %s", cnpName)
 			continue
 		}
-		n.reprocessCNP(cnp)
+		n.reprocessCNP(cnp, false)
 	}
 }
 

--- a/pkg/controller/networkpolicy/networkpolicy_controller.go
+++ b/pkg/controller/networkpolicy/networkpolicy_controller.go
@@ -215,6 +215,75 @@ type heartbeat struct {
 	timestamp time.Time
 }
 
+var tierIndexers = cache.Indexers{
+	PriorityIndex: func(obj interface{}) ([]string, error) {
+		tr, ok := obj.(*secv1alpha1.Tier)
+		if !ok {
+			return []string{}, nil
+		}
+		return []string{strconv.FormatInt(int64(tr.Spec.Priority), 10)}, nil
+	},
+}
+
+var cnpIndexers = cache.Indexers{
+	TierIndex: func(obj interface{}) ([]string, error) {
+		cnp, ok := obj.(*secv1alpha1.ClusterNetworkPolicy)
+		if !ok {
+			return []string{}, nil
+		}
+		return []string{cnp.Spec.Tier}, nil
+	},
+	ClusterGroupIndex: func(obj interface{}) ([]string, error) {
+		cnp, ok := obj.(*secv1alpha1.ClusterNetworkPolicy)
+		if !ok {
+			return []string{}, nil
+		}
+		groupNames := sets.String{}
+		for _, appTo := range cnp.Spec.AppliedTo {
+			if appTo.Group != "" {
+				groupNames.Insert(appTo.Group)
+			}
+		}
+		if len(cnp.Spec.Ingress) == 0 && len(cnp.Spec.Egress) == 0 {
+			return groupNames.List(), nil
+		}
+		appendGroups := func(rule secv1alpha1.Rule) {
+			for _, peer := range rule.To {
+				if peer.Group != "" {
+					groupNames.Insert(peer.Group)
+				}
+			}
+			for _, peer := range rule.From {
+				if peer.Group != "" {
+					groupNames.Insert(peer.Group)
+				}
+			}
+			for _, appTo := range rule.AppliedTo {
+				if appTo.Group != "" {
+					groupNames.Insert(appTo.Group)
+				}
+			}
+		}
+		for _, rule := range cnp.Spec.Egress {
+			appendGroups(rule)
+		}
+		for _, rule := range cnp.Spec.Ingress {
+			appendGroups(rule)
+		}
+		return groupNames.List(), nil
+	},
+}
+
+var anpIndexers = cache.Indexers{
+	TierIndex: func(obj interface{}) ([]string, error) {
+		anp, ok := obj.(*secv1alpha1.NetworkPolicy)
+		if !ok {
+			return []string{}, nil
+		}
+		return []string{anp.Spec.Tier}, nil
+	},
+}
+
 // NewNetworkPolicyController returns a new *NetworkPolicyController.
 func NewNetworkPolicyController(kubeClient clientset.Interface,
 	crdClient versioned.Interface,
@@ -296,67 +365,8 @@ func NewNetworkPolicyController(kubeClient clientset.Interface,
 			},
 			resyncPeriod,
 		)
-		tierInformer.Informer().AddIndexers(
-			cache.Indexers{
-				PriorityIndex: func(obj interface{}) ([]string, error) {
-					tr, ok := obj.(*secv1alpha1.Tier)
-					if !ok {
-						return []string{}, nil
-					}
-					return []string{strconv.FormatInt(int64(tr.Spec.Priority), 10)}, nil
-				},
-			},
-		)
-		cnpInformer.Informer().AddIndexers(
-			cache.Indexers{
-				TierIndex: func(obj interface{}) ([]string, error) {
-					cnp, ok := obj.(*secv1alpha1.ClusterNetworkPolicy)
-					if !ok {
-						return []string{}, nil
-					}
-					return []string{cnp.Spec.Tier}, nil
-				},
-				ClusterGroupIndex: func(obj interface{}) ([]string, error) {
-					cnp, ok := obj.(*secv1alpha1.ClusterNetworkPolicy)
-					if !ok {
-						return []string{}, nil
-					}
-					groupNames := sets.String{}
-					for _, appTo := range cnp.Spec.AppliedTo {
-						if appTo.Group != "" {
-							groupNames.Insert(appTo.Group)
-						}
-					}
-					if len(cnp.Spec.Ingress) == 0 && len(cnp.Spec.Egress) == 0 {
-						return groupNames.List(), nil
-					}
-					appendGroups := func(rule secv1alpha1.Rule) {
-						for _, peer := range rule.To {
-							if peer.Group != "" {
-								groupNames.Insert(peer.Group)
-							}
-						}
-						for _, peer := range rule.From {
-							if peer.Group != "" {
-								groupNames.Insert(peer.Group)
-							}
-						}
-						for _, appTo := range rule.AppliedTo {
-							if appTo.Group != "" {
-								groupNames.Insert(appTo.Group)
-							}
-						}
-					}
-					for _, rule := range cnp.Spec.Egress {
-						appendGroups(rule)
-					}
-					for _, rule := range cnp.Spec.Ingress {
-						appendGroups(rule)
-					}
-					return groupNames.List(), nil
-				},
-			},
-		)
+		tierInformer.Informer().AddIndexers(tierIndexers)
+		cnpInformer.Informer().AddIndexers(cnpIndexers)
 		cnpInformer.Informer().AddEventHandlerWithResyncPeriod(
 			cache.ResourceEventHandlerFuncs{
 				AddFunc:    n.addCNP,
@@ -365,17 +375,7 @@ func NewNetworkPolicyController(kubeClient clientset.Interface,
 			},
 			resyncPeriod,
 		)
-		anpInformer.Informer().AddIndexers(
-			cache.Indexers{
-				TierIndex: func(obj interface{}) ([]string, error) {
-					anp, ok := obj.(*secv1alpha1.NetworkPolicy)
-					if !ok {
-						return []string{}, nil
-					}
-					return []string{anp.Spec.Tier}, nil
-				},
-			},
-		)
+		anpInformer.Informer().AddIndexers(anpIndexers)
 		anpInformer.Informer().AddEventHandlerWithResyncPeriod(
 			cache.ResourceEventHandlerFuncs{
 				AddFunc:    n.addANP,

--- a/pkg/controller/networkpolicy/networkpolicy_controller_perf_test.go
+++ b/pkg/controller/networkpolicy/networkpolicy_controller_perf_test.go
@@ -471,7 +471,7 @@ func benchmarkInit(b *testing.B, namespaces []*corev1.Namespace, networkPolicies
 	objs := toRunTimeObjects(namespaces, networkPolicies, pods)
 	stopCh := make(chan struct{})
 	defer close(stopCh)
-	_, c := newControllerWithoutEventHandler(objs...)
+	_, c := newControllerWithoutEventHandler(objs, nil)
 	c.informerFactory.Start(stopCh)
 	c.informerFactory.WaitForCacheSync(stopCh)
 

--- a/pkg/controller/networkpolicy/networkpolicy_controller_test.go
+++ b/pkg/controller/networkpolicy/networkpolicy_controller_test.go
@@ -143,9 +143,9 @@ func newController(objects ...runtime.Object) (*fake.Clientset, *networkPolicyCo
 
 // newControllerWithoutEventHandler creates a networkPolicyController that doesn't register event handlers so that the
 // tests can call event handlers in their own ways.
-func newControllerWithoutEventHandler(objects ...runtime.Object) (*fake.Clientset, *networkPolicyController) {
-	client := newClientset(objects...)
-	crdClient := fakeversioned.NewSimpleClientset()
+func newControllerWithoutEventHandler(k8sObjects, crdObjects []runtime.Object) (*fake.Clientset, *networkPolicyController) {
+	client := newClientset(k8sObjects...)
+	crdClient := fakeversioned.NewSimpleClientset(crdObjects...)
 	informerFactory := informers.NewSharedInformerFactory(client, informerDefaultResync)
 	crdInformerFactory := crdinformers.NewSharedInformerFactory(crdClient, informerDefaultResync)
 	appliedToGroupStore := store.NewAppliedToGroupStore()
@@ -153,7 +153,10 @@ func newControllerWithoutEventHandler(objects ...runtime.Object) (*fake.Clientse
 	internalNetworkPolicyStore := store.NewNetworkPolicyStore()
 	internalGroupStore := store.NewGroupStore()
 	networkPolicyInformer := informerFactory.Networking().V1().NetworkPolicies()
-	cgStore := crdInformerFactory.Crd().V1alpha2().ClusterGroups().Informer().GetStore()
+	tierInformer := crdInformerFactory.Crd().V1alpha1().Tiers()
+	cnpInformer := crdInformerFactory.Crd().V1alpha1().ClusterNetworkPolicies()
+	anpInformer := crdInformerFactory.Crd().V1alpha1().NetworkPolicies()
+	cgInformer := crdInformerFactory.Crd().V1alpha3().ClusterGroups()
 	groupEntityIndex := grouping.NewGroupEntityIndex()
 	npController := &NetworkPolicyController{
 		kubeClient:                 client,
@@ -161,6 +164,18 @@ func newControllerWithoutEventHandler(objects ...runtime.Object) (*fake.Clientse
 		networkPolicyInformer:      networkPolicyInformer,
 		networkPolicyLister:        networkPolicyInformer.Lister(),
 		networkPolicyListerSynced:  networkPolicyInformer.Informer().HasSynced,
+		tierInformer:               tierInformer,
+		tierLister:                 tierInformer.Lister(),
+		tierListerSynced:           tierInformer.Informer().HasSynced,
+		cnpInformer:                cnpInformer,
+		cnpLister:                  cnpInformer.Lister(),
+		cnpListerSynced:            cnpInformer.Informer().HasSynced,
+		anpInformer:                anpInformer,
+		anpLister:                  anpInformer.Lister(),
+		anpListerSynced:            anpInformer.Informer().HasSynced,
+		cgInformer:                 cgInformer,
+		cgLister:                   cgInformer.Lister(),
+		cgListerSynced:             cgInformer.Informer().HasSynced,
 		addressGroupStore:          addressGroupStore,
 		appliedToGroupStore:        appliedToGroupStore,
 		internalNetworkPolicyStore: internalNetworkPolicyStore,
@@ -171,14 +186,17 @@ func newControllerWithoutEventHandler(objects ...runtime.Object) (*fake.Clientse
 		internalGroupQueue:         workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "internalGroup"),
 		groupingInterface:          groupEntityIndex,
 	}
+	npController.tierInformer.Informer().AddIndexers(tierIndexers)
+	npController.cnpInformer.Informer().AddIndexers(cnpIndexers)
+	npController.anpInformer.Informer().AddIndexers(anpIndexers)
 	return client, &networkPolicyController{
 		npController,
 		informerFactory.Core().V1().Namespaces().Informer().GetStore(),
 		informerFactory.Core().V1().Services().Informer().GetStore(),
 		informerFactory.Networking().V1().NetworkPolicies().Informer().GetStore(),
-		crdInformerFactory.Crd().V1alpha1().ClusterNetworkPolicies().Informer().GetStore(),
-		crdInformerFactory.Crd().V1alpha1().Tiers().Informer().GetStore(),
-		cgStore,
+		cnpInformer.Informer().GetStore(),
+		tierInformer.Informer().GetStore(),
+		cgInformer.Informer().GetStore(),
 		appliedToGroupStore,
 		addressGroupStore,
 		internalNetworkPolicyStore,

